src/sequence_batch_scheduler.cc

```diff
@@ -798,85 +835,140 @@ SequenceBatchScheduler::ReaperThread(const int nice)
+            // If the idle correlation ID is in the backlog, then just
+            // need to increase the timeout so that we revisit it again in
+            // the future to check if it is assigned to a sequence slot.
+            auto idle_bl_itr =
+                sequence_to_backlog_map_.find(idle_correlation_id);
+            if (idle_bl_itr != sequence_to_backlog_map_.end()) {
+              LOG_VERBOSE(1)
+                  << "Reaper: found idle CORRID " << idle_correlation_id;
+              wait_microseconds =
+                  std::min(wait_microseconds, backlog_idle_wait_microseconds);
+              ++cid_itr;
+            } else {
+              LOG_VERBOSE(1) << "Reaper: ignoring stale idle CORRID "
+                             << idle_correlation_id;
+              cid_itr = correlation_id_timestamps_.erase(cid_itr);
+            }
           }
         }
       }
+
+      // Enqueue force-ends outside of the lock.
+      for (const auto& pr : force_end_sequences) {
+        const InferenceRequest::SequenceId& idle_correlation_id = pr.first;
+        const size_t batcher_idx = pr.second.batcher_idx_;
+        const uint32_t seq_slot = pr.second.seq_slot_;
+
+        LOG_VERBOSE(1) << "Reaper: force-ending CORRID " << idle_correlation_id
+                       << " in batcher " << batcher_idx << ", slot "
+                       << seq_slot;
+
+        // A slot assignment is released by enqueuing a request with a
+        // null request. The scheduler thread will interpret the null
+        // request as meaning it should release the sequence slot but
+        // otherwise do nothing with the request.
+        std::unique_ptr<InferenceRequest> null_request;
+        batchers_[batcher_idx]->Enqueue(
+            seq_slot, idle_correlation_id, null_request);
+      }
+
+      // Update timestamp for next idle check
+      idle_timestamp = now_us + wait_microseconds;
     }
 
-    // Enqueue force-ends outside of the lock.
-    for (const auto& pr : force_end_sequences) {
-      const InferenceRequest::SequenceId& idle_correlation_id = pr.first;
-      const size_t batcher_idx = pr.second.batcher_idx_;
-      const uint32_t seq_slot = pr.second.seq_slot_;
-
-      LOG_VERBOSE(1) << "Reaper: force-ending CORRID " << idle_correlation_id
-                     << " in batcher " << batcher_idx << ", slot " << seq_slot;
-
-      // A slot assignment is released by enqueuing a request with a
-      // null request. The scheduler thread will interpret the null
-      // request as meaning it should release the sequence slot but
-      // otherwise do nothing with the request.
-      std::unique_ptr<InferenceRequest> null_request;
-      batchers_[batcher_idx]->Enqueue(
-          seq_slot, idle_correlation_id, null_request);
+    // Reap timed out backlog sequence
+    if (now_us >= timeout_timestamp_) {
+      timeout_timestamp_ = std::numeric_limits<uint64_t>::max();
+      std::deque<std::shared_ptr<BacklogQueue>> expired_backlogs;
```

Comment:
Would using a vector be preferable here? You don't seem to be using any properties where deques are superior here (e.g. insertion/removal at the start), so the additional bookkeeping of deques (e.g. associated with them being stored non-contiguously) seems like unnecessary overhead.

Reply:
I don't think the container type matters too much here. Using vector does help when iterating the element given the contiguity, but rejecting the request is done at the end of reaper loop and the mutex is released so the performance is not critical. Another thing for using vector is that there can be resizing of the container on insertion and causing overhead.